pkg/flare/service_windows.go

```diff
@@ -0,0 +1,286 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build windows
+
+package flare
+
+import (
+	"fmt"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+const (
+	scManagerAccess = windows.SC_MANAGER_CONNECT | windows.SC_MANAGER_ENUMERATE_SERVICE
+)
+
+type serviceInfo struct {
+	ServiceName           string
+	ServiceState          string
+	ProcessID             uint32
+	Config                queryServiceConfig
+	DependentServices     []string
+	TriggersCount         uint32
+	ServiceFailureActions serviceFailureActions
+}
+
+type queryServiceConfig struct {
+	ServiceType      string
+	StartType        string
+	ErrorControl     string
+	BinaryPathName   string // fully qualified path to the service binary file, can also include arguments for an auto-start service
+	LoadOrderGroup   string
+	TagID            uint32
+	Dependencies     []string
+	ServiceStartName string // name of the account under which the service should run
+	DisplayName      string
+	Password         string
+	Description      string
+	SidType          uint32 // one of SERVICE_SID_TYPE, the type of sid to use for the service
+	DelayedAutoStart bool   // the service is started after other auto-start services are started plus a short delay
+}
+
+type recoveryActionsUpdated struct {
+	Type  string
+	Delay time.Duration
+}
+
+type serviceFailureActions struct {
+	ResetPeriod                      uint32
+	RebootMessage                    string
+	RecoveryCommand                  string
+	FailureActionsOnNonCrashFailures bool
+	RecoveryActions                  []recoveryActionsUpdated
+}
+
+func getFailureActions(s *mgr.Service) (serviceFailureActions, error) {
+	sfa := serviceFailureActions{}
+
+	var err error
+	var err1 error
+
+	sfa.ResetPeriod, err1 = s.ResetPeriod()
+	if err1 != nil {
+		err = fmt.Errorf("%v, ResetPeriod: %v", err, err1)
+	}
+
+	sfa.RebootMessage, err1 = s.RebootMessage()
+	if err1 != nil {
+		err = fmt.Errorf("%v, RebootMessage: %v", err, err1)
+	}
+
+	sfa.RecoveryCommand, err1 = s.RecoveryCommand()
+	if err1 != nil {
+		err = fmt.Errorf("%v, RecoveryCommand: %v", err, err1)
+	}
+
+	sfa.FailureActionsOnNonCrashFailures, err1 = s.RecoveryActionsOnNonCrashFailures()
+	if err1 != nil {
+		err = fmt.Errorf("%v, FailureActionsOnNonCrashFailure: %v", err, err1)
+	}
+
+	recoveryActionSlice, err1 := s.RecoveryActions()
+	if err1 != nil {
+		err = fmt.Errorf("%v, RecoveryActions: %v", err, err1)
+	}
+
+	for _, act := range recoveryActionSlice {
```

Comment:
Why we just not copy return to sfa.RecoveryActions? What is the point of dividing original Delay?

Reply:
Hi @iglendd , 

With `s.RecoveryActions()` the following struct is returned: https://learn.microsoft.com/en-us/windows/win32/api/winsvc/ns-winsvc-sc_action

The original struct from Microsoft has a numerical value for `SC_ACTION_TYPE` which needed to be converted to a string that would be more readable for a person viewing this information. Which is why we needed to parse through `recoveryActionSlice` to convert that number to a string.

For some reason the GO windows service manger package multiplies the value for Delay returned by Microsoft by `time.Milliseconds`: https://cs.opensource.google/go/x/sys/+/refs/tags/v0.12.0:windows/svc/mgr/recovery.go;l=81

I had to reverse that because Microsoft already returns the Delay in milliseconds so the multiplication was unnecessary on GO's part.

Reply:
Interesting, thank you makes sense.